### PR TITLE
Epic merge: ipfs-cluster-dev and DSL features

### PR DESCRIPTION
**Pull request by ZenGround0**
@hsanjuan @dgrisham This is the merge of cluster-dev and the two branches I have going for DSL features.  Those branches each have their own well documented PRs #23 and #25 .  Weak spots include testing integration, I have yet to get main_test.go running the tests for 3 of the 4 mods of #25 and the tests for these mods really require manual inspection of files right now,  and validation.  I have only added validation for the fourth modification of #25 and perhaps it will be more user-friendly to do some validation involving the `for` and `append_on` structures and maybe even with command strings.  The regexes I use work with decent generality but perhaps we can do better.  The environment's inclusion of arrays is a bit sloppy (every run that uses the array appends the whole thing to env and never takes it off), so I am interested in your suggestions for any improvements here too.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -110,25 +230,60 @@ func main() {
...
-	fileData, err := ioutil.ReadFile(filePath)
+	var subsetPartition map[int][]int
+	err, test := readTestFile(filePath)
 	if err != nil {
 		fatal(err)
 	}
 
-	var test Test
-	var summary Summary
+	rand.Seed(time.Now().UTC().UnixNano())
```
Comment (hsanjuan) [inline main.go]: yeah better put this inside an `init()` function


https://medium.com/golangspec/init-functions-in-go-eac191b3860a

File: main.go [hunk 2]
```diff
@@ -110,25 +230,60 @@ func main() {
 	filePath := os.Args[1]
 	debug("## Loading " + filePath)
 
-	fileData, err := ioutil.ReadFile(filePath)
+	var subsetPartition map[int][]int
+	err, test := readTestFile(filePath)
```
Comment (hsanjuan) [inline main.go]: `test, err :=`.  I have never seen a case where err is the first thing returned by a function

File: main.go [hunk 3]
```diff
@@ -110,25 +230,60 @@ func main() {
 	filePath := os.Args[1]
 	debug("## Loading " + filePath)
 
-	fileData, err := ioutil.ReadFile(filePath)
+	var subsetPartition map[int][]int
```
Comment (hsanjuan) [inline main.go]: this line can go

File: main.go [hunk 4]
```diff
@@ -110,25 +230,60 @@ func main() {
...
+	err, test := readTestFile(filePath)
 	if err != nil {
 		fatal(err)
 	}
 
-	var test Test
-	var summary Summary
+	rand.Seed(time.Now().UTC().UnixNano())
+	err2, subsetPartition := partition(test.Config)
```
Comment (hsanjuan) [inline main.go]: `subsetParititon, err := `. You should be able to re-use `err`.

**Comment by ZenGround0** [comment]
@hsanjuan Ok I have addressed this feedback.  Get back to me with any other thoughts.

**Comment by ZenGround0** [comment]
@hsanjuan These new changes fix up the sloppy inclusion of arrays mentioned in the environment along with the reordering of selection so that it happens every iteration allowing iterations to proceed on random nodes.

**Comment by hsanjuan** [comment]
You will need to rebase probably before merge

**Comment by ZenGround0** [comment]
Cool, tests run successfully with these changes.  I am ready to rebase once the parameters PR is merged in.

**Pull request merged**